### ISSUE TITLE
Bug: thread 'main' panicked at 'attempt to shift left with overflow', src/mempool.rs:138:25

### ISSUE THREAD

**Issue by yojoe** [romanz/electrs#728]
**Describe the bug**
`electrs` debug builds panic when a block is mined on `regtest`. Only happens with **debug** builds and not with **release** builds:

```
INFO  electrs::index] indexing 101 blocks: [1..101]
INFO  electrs::chain] chain updated: tip=0471ac0d97f003da4446f65eb0d0944745de958b399c60f8bcfbad9584444527, height=101
thread 'main' panicked at 'attempt to shift left with overflow', src/mempool.rs:138:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
INFO  electrs::db] closing DB at ./db/regtest
```
https://github.com/romanz/electrs/blob/d2581c77f794894281de2ab64f987963d31af90a/src/mempool.rs#L138

**Electrs version**
v0.9.8 (debug build) aarch64

**To Reproduce**
1. Create a `debug` build: `cargo build`.  (No release build!!!)
2. Start `bitcoind` in regtest.
3. Start `electrs` in regtest.
4. Mine at least one block. (Example: `bitcoin-cli -regtest generatetoaddress 101 bcrt1qf32trg48tw6emq355u4x7ktaqvuatesafvwkcj`)
5. See error

**Expected behavior**
No panic / overflow in `src/mempool.rs:138:25` for **debug** builds.

**System running electrs**
 - Deployment method: manual native OS
 - OS name and version: Debian 11 (aarch64)

**Additional context**
Similar Rust issue which causes panics on **debug** builds: https://users.rust-lang.org/t/intentionally-overflow-on-shift/11859

**Comment by engenegr** [comment]
See it again on v0.9.9 for debug build.

**Comment by engenegr** [comment]
Not seeing it on release builds.

**Comment by romanz** [comment]
Could you please attach the error log + full backtrace?

**Comment by engenegr** [comment]
Oh, I did even remove whole repository and moved on Blockstream version since I needed Esplora HTTP for integration tests. I will do it again, i think.

**Comment by engenegr** [comment]
Full dump

```
[2022-12-28T10:48:23.432Z INFO  electrs::server] serving Electrum RPC on 127.0.0.1:18448
[2022-12-28T10:48:23.432Z DEBUG tiny_http] Server listening on 127.0.0.1:24224
[2022-12-28T10:48:23.433Z DEBUG tiny_http] Running accept thread
[2022-12-28T10:48:23.440Z INFO  electrs::db] "/tmp/regtest-electrs-data.4Ve84nFRITXg/regtest": 7 SST files, 0.000129589 GB, 0.000004532 Grows
[2022-12-28T10:48:23.440Z DEBUG electrs::db] DB Some(Config { compacted: true, format: 0 })
[2022-12-28T10:48:23.451Z DEBUG electrs::db] auto-compactions enabled
[2022-12-28T10:48:23.487Z INFO  electrs::chain] loading 1246 headers, tip=672c7536541abe3979fb2529a0fba9fb5701998965c08e6bfdf3287768cb40f6
[2022-12-28T10:48:23.544Z INFO  electrs::chain] chain updated: tip=672c7536541abe3979fb2529a0fba9fb5701998965c08e6bfdf3287768cb40f6, height=1246
[2022-12-28T10:48:23.548Z DEBUG bitcoincore_rpc] JSON-RPC request: getblockchaininfo []
[2022-12-28T10:48:23.549Z DEBUG bitcoincore_rpc] JSON-RPC request: getnetworkinfo []
[2022-12-28T10:48:23.549Z DEBUG bitcoincore_rpc] JSON-RPC request: getnetworkinfo []
[2022-12-28T10:48:23.550Z DEBUG bitcoincore_rpc] JSON-RPC request: getblockchaininfo []
[2022-12-28T10:48:23.551Z DEBUG bitcoincore_rpc] JSON-RPC request: getnetworkinfo []
[2022-12-28T10:48:23.553Z DEBUG electrs::p2p] peer version: VersionMessage { version: 70016, services: ServiceFlags(1033), timestamp: 1672224503, receiver: Address {services: ServiceFlags(NONE), address: 0.0.0.0, port: 0}, sender: Address {services: ServiceFlags(NETWORK|WITNESS|NETWORK_LIMITED), address: 0.0.0.0, port: 0}, nonce: 13135438854695354256, user_agent: "/Satoshi:22.0.0/", start_height: 1246, relay: true }
[2022-12-28T10:48:23.555Z DEBUG electrs::p2p] got 0 new headers
[2022-12-28T10:48:23.557Z DEBUG bitcoincore_rpc] JSON-RPC request: getrawmempool []
[2022-12-28T10:48:23.557Z DEBUG electrs::mempool] loading 0 mempool transactions
thread 'main' panicked at 'attempt to shift left with overflow', src/mempool.rs:138:25
stack backtrace:
   0:     0x55b80ab25400 - std::backtrace_rs::backtrace::libunwind::trace::h72c2fb8038f1bbee
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/../../backtrace/src/backtrace/libunwind.rs:96
   1:     0x55b80ab25400 - std::backtrace_rs::backtrace::trace_unsynchronized::h1e3b084883f1e78c
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/../../backtrace/src/backtrace/mod.rs:66
   2:     0x55b80ab25400 - std::sys_common::backtrace::_print_fmt::h3bf6a7ebf7f0394a
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:79
   3:     0x55b80ab25400 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h2e8cb764b7fe02e7
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:58
   4:     0x55b80ab48c4c - core::fmt::write::h7a1184eaee6a8644
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/fmt/mod.rs:1080
   5:     0x55b80ab1dd02 - std::io::Write::write_fmt::haeeb374d93a67eac
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/io/mod.rs:1516
   6:     0x55b80ab2780d - std::sys_common::backtrace::_print::h1d14a7f6ad632dc8
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:61
   7:     0x55b80ab2780d - std::sys_common::backtrace::print::h301abac8bb2e3e81
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:48
   8:     0x55b80ab2780d - std::panicking::default_hook::{{closure}}::hde0cb80358a6920a
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:208
   9:     0x55b80ab274b8 - std::panicking::default_hook::h9b1a691049a0ec8f
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:227
  10:     0x55b80ab27ef1 - std::panicking::rust_panic_with_hook::h2bdec87b60580584
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:577
  11:     0x55b80ab27a99 - std::panicking::begin_panic_handler::{{closure}}::h101ca09d9df5db47
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:484
  12:     0x55b80ab2586c - std::sys_common::backtrace::__rust_end_short_backtrace::h3bb85654c20113ca
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:153
  13:     0x55b80ab27a59 - rust_begin_unwind
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:483
  14:     0x55b80ab46831 - core::panicking::panic_fmt::h48c31e1e3d550146
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/panicking.rs:85
  15:     0x55b80ab4677d - core::panicking::panic::h184ede6dd822ffb4
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/panicking.rs:50
  16:     0x55b809ce9062 - electrs::mempool::Mempool::sync::h61d2cd9830f3562f
                               at /tmp/electrs/src/mempool.rs:138
  17:     0x55b809cbafae - electrs::tracker::Tracker::sync::h65735c0d8178e529
                               at /tmp/electrs/src/tracker.rs:68
  18:     0x55b809d86595 - electrs::electrum::Rpc::sync::h4c82d86dc108a747
                               at /tmp/electrs/src/electrum.rs:164
  19:     0x55b809c096be - electrs::server::serve::{{closure}}::hbc1d02b7d697d4ed
                               at /tmp/electrs/src/server.rs:93
  20:     0x55b809fc53ac - prometheus::histogram::Histogram::observe_closure_duration::h912f595cb02cd9ff
                               at /home/ilya/.cargo/registry/src/github.com-1ecc6299db9ec823/prometheus-0.13.0/src/histogram.rs:716
  21:     0x55b809d39a7c - electrs::metrics::metrics_impl::Histogram::observe_duration::hd8dd8c6f244f054b
                               at /tmp/electrs/src/metrics.rs:97
  22:     0x55b809bcee73 - electrs::server::serve::h5339c3cf23fc3d07
                               at /tmp/electrs/src/server.rs:93
  23:     0x55b809bce05c - electrs::server::run::h54a26073447c54ac
                               at /tmp/electrs/src/server.rs:51
  24:     0x55b809bcd4d7 - electrs::main::h3dbf0404bce56209
                               at /tmp/electrs/src/bin/electrs.rs:6
  25:     0x55b809bcd38b - core::ops::function::FnOnce::call_once::h3f17187b466672de
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/ops/function.rs:227
  26:     0x55b809bcd52e - std::sys_common::backtrace::__rust_begin_short_backtrace::h1991593e1d3dc656
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:137
  27:     0x55b809bcd461 - std::rt::lang_start::{{closure}}::hdc868d48c211f0d0
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/rt.rs:66
  28:     0x55b80ab28407 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::he179d32a5d10d957
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/ops/function.rs:259
  29:     0x55b80ab28407 - std::panicking::try::do_call::hcb3d5e7be089b2b4
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:381
  30:     0x55b80ab28407 - std::panicking::try::h7ac93b0cd56fb701
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:345
  31:     0x55b80ab28407 - std::panic::catch_unwind::h7b40e396c93a4fcd
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panic.rs:382
  32:     0x55b80ab28407 - std::rt::lang_start_internal::h142b9cc66267fea1
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/rt.rs:51
  33:     0x55b809bcd437 - std::rt::lang_start::h2ffd82ed864fd2e8
                               at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/rt.rs:65
  34:     0x55b809bcd51a - main
  35:     0x7f3f59629510 - __libc_start_call_main
  36:     0x7f3f596295c9 - __libc_start_main@GLIBC_2.2.5
  37:     0x55b809bcd0c5 - _start
  38:                0x0 - <unknown>
[2022-12-28T10:48:23.605Z INFO  electrs::db] closing DB at /tmp/regtest-electrs-data.4Ve84nFRITXg/regtest
[2022-12-28T10:48:23.606Z DEBUG electrs::p2p] closing p2p_loop thread: no more requests to handle
[2022-12-28T10:48:23.606Z DEBUG electrs::p2p] closing p2p_send thread: no more messages to send
```

**Comment by romanz** [comment]
It should be fixed in v0.9.10:
https://github.com/romanz/electrs/commit/1592983dfd78c2ff562d8bab41fc2fd60a8af901
